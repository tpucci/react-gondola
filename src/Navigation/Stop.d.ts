import { ComponentType } from 'react';

export interface IStop {
  Component: ComponentType<IStopComponentProps>;
  isFullScreen?: boolean;
  name: string;
  isAuthorized: boolean;
}

export interface IStopComponentProps {
  isAuthorized: boolean;
}
