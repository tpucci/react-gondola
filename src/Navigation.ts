import { Subject, Observable } from 'rxjs';
import { ComponentType } from 'react';
import { mergeAll, map, scan } from 'rxjs/operators';

export interface IStop {
  name: string;
  Component: ComponentType;
  isFullScreen?: boolean;
}

type Stack = IStop[];

interface IFullScreenStackProperties {
  canalId: string;
  fullScreenStack: Stack;
}

interface IFullScreenStackMap {
  [canalId: string]: Stack;
}

export class Navigation {
  static getInstance() {
    if (Navigation.instance) {
      return Navigation.instance;
    }
    Navigation.instance = new Navigation();
    return Navigation.instance;
  }
  private static instance: Navigation;

  canalsFullScreenStackProperties$ = new Subject<
    Observable<IFullScreenStackProperties>
  >();

  fullSceenStack$: Observable<
    Stack
  > = this.canalsFullScreenStackProperties$.pipe(
    mergeAll(),
    scan(
      (
        fullScreenStackMap: IFullScreenStackMap,
        fullScreenStackProperties: IFullScreenStackProperties
      ) => ({
        ...fullScreenStackMap,
        [fullScreenStackProperties.canalId]:
          fullScreenStackProperties.fullScreenStack
      }),
      {}
    ),
    map(fullScreenStackMap =>
      Object.keys(fullScreenStackMap).reduce(
        (fullScreenStack: Stack, canalId: string) =>
          fullScreenStack.concat(fullScreenStackMap[canalId]),
        []
      )
    )
  );
}
