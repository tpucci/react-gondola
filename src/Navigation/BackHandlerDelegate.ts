import { BackHandler } from 'react-native';
import { fromEventPattern, Observable, Subject } from 'rxjs';
import { share, map, merge, tap } from 'rxjs/operators';

export interface BackEvent {
  target: string | null;
}

export class BackHandlerDelegate {
  /**
   * Attach the given handler to the event source. When the source emits,
   * calls the handler. The handler is responsible to register a callback.
   * If it succeeded to register a callback, we call it.
   * If not, and no other react native back event listener has been registered
   * elsewhere, this listener returns false which might kill the app.
   * See https://facebook.github.io/react-native/docs/backhandler.html.
   * @param handler
   */
  private addEventListener = (handler: () => void) => {
    BackHandler.addEventListener('hardwareBackPress', () => {
      handler();
      return true;
    });
  };

  private softwareBack$ = new Subject<undefined>();

  private back$: Observable<BackEvent> = fromEventPattern<undefined>(this.addEventListener).pipe(
    merge(this.softwareBack$),
    tap(() => {
      // After the event has spread, we fire onBacks callbacks.
      setTimeout(() => {
        if (this.onBackCallback) {
          this.onBackCallback();
          this.onBackCallback = undefined;
        } else {
          BackHandler.exitApp();
        }
      }, 0);
    }),
    map(() => ({
      target: null,
    })),
    share()
  );

  private onBackCallback?: () => any;
  setOnBackCallback = (cb: () => any) => {
    this.onBackCallback = cb;
  };

  back = () => {
    this.softwareBack$.next();
  };

  defaultBackContext = { back$: this.back$ };
}
