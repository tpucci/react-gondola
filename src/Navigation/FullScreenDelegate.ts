import { Subject, Observable } from 'rxjs';
import { map, scan, tap } from 'rxjs/operators';
import { ReactElement } from 'react';
import { ScreenProps, Screen } from '../Screen';

type Stack = ReactElement<ScreenProps, typeof Screen>[];

interface FullScreenStackProperties {
  canalId: string;
  fullScreenStack: Stack;
}

interface FullScreenStackMap {
  [canalId: string]: Stack;
}

export class FullScreenDelegate {
  canalsFullScreenStackProperties$ = new Subject<FullScreenStackProperties>();

  fullSceenStack$: Observable<Stack> = this.canalsFullScreenStackProperties$.pipe(
    scan(
      (
        fullScreenStackMap: FullScreenStackMap,
        fullScreenStackProperties: FullScreenStackProperties
      ) => ({
        ...fullScreenStackMap,
        [fullScreenStackProperties.canalId]: fullScreenStackProperties.fullScreenStack,
      }),
      {}
    ),
    map((fullScreenStackMap: FullScreenStackMap) =>
      Object.keys(fullScreenStackMap).reduce(
        (fullScreenStack: Stack, canalId: string) =>
          fullScreenStack.concat(fullScreenStackMap[canalId]),
        []
      )
    ),
    tap(o => console.warn(o))
  );
}
