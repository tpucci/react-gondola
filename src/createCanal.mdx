---
name: createCanal
menu: API
route: /api/create-canal
---

# createCanal

`type: <Authorizations = { [K in Stop['name']]?: boolean }>(stopsList: Stop[]): ComponentType<CanalComponentProps<Authorizations>>`

`createCanal` takes a list of `Stop` and returns a new canal. A valid `Stop` is an object with a unique `name` key, a component key and an optional `isFullScreen` flag. See `Stop` for more details.

## Stop

|    Key name    | Required | Default     | Usage                                                                           |
| :------------: | :------: | ----------- | ------------------------------------------------------------------------------- |
|     `name`     |  `true`  | `undefined` | Unique identifier.                                                              |
|  `Component`   |  `true`  | `undefined` | The component that will be rendered when the user progress reaches that stop.   |
|    `props`     | `false`  | `undefined` | Optional. Props that will be passed down to the component.                      |
|    `onBack`    | `false`  | `undefined` | Optional. Android Back button Callback.                                         |
| `isFullScreen` | `false`  | `undefined` | Optional. Indicates if the stop needs to be rendered in the `FullScreenPortal`. |

## Canal Props

|       Prop name       | Required | Default                               | Usage                                                                           |
| :-------------------: | :------: | ------------------------------------- | ------------------------------------------------------------------------------- |
|        `style`        | `false`  | `ReactNative.StyleSheet.absoluteFill` | Style applied to the Canal wrapper.                                             |
| `[K in Stop['name']]` | `false`  | `undefined`                           | Indicates to the canal that we are authorized to render the corresponding stop. |

## Canal Properties

### `canalId: string`

The canal unique id generated with `Date.now().toString()`

### `stopsList: Stop[]`

The list of Stop given as the `createCanal` first argument.

### `authorizations: Subject<Authorizations>`

RxJS Subject of Authorizations.

### `progress: Observable<Stop[]>`

RxJS Observable of the list of authorized stops which represent a valid route. A valid route starts from the first stop, go through each stop and ends at the first non-authorized stop. If the first stop is not authorized, the progress emits an empty list.

### `stack: IStreamListener<Stop[]>`

MobX Strem Listener of list of local Stops. All full screen stops are filtered out. This stream is observed by MobX and fires the rendering of the canal stack in the `Canal.render` method.

## Basic usage

Link your pages in a canal.

```tsx
import React from "react";
import { createCanal } from "react-gondola";
import { FirstName } from "./FirstName";
import { LastName } from "./LastName";
import { Confirm } from "./Confirm";

export const SignInCanal = createCanal([
  { name: "firstName", Component: FirstName },
  { name: "lastName", Component: LastName },
  { name: "confirm", Component: Confirm, isFullScreen: true }
]);
```
