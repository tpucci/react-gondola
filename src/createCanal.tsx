import React, { ComponentType, Component as ReactComponent } from 'react';
import { ViewStyle, View, StyleSheet, StyleProp } from 'react-native';
import { observer } from 'mobx-react/native';
import { Canal, IStop } from './Canal';
import { Navigation } from './Navigation.store';

type CanalComponentProps<T> = {
  style?: StyleProp<ViewStyle>;
} & T;

export const createCanal = <
  T extends string,
  U extends { name: T } & IStop,
  V = { [K in U['name']]?: boolean }
>(
  ...StopsList: U[]
): ComponentType<CanalComponentProps<V>> => {
  for (let index = 0; index < StopsList.length; index++) {
    const Stop = StopsList[index];
    if (!Stop.name || typeof Stop.name !== 'string') {
      throw new Error(
        `\`createCanal\` could not find a valid \`name\` key for argument ${index +
          1}. Received: ${JSON.stringify(Stop)}`
      );
    }
    if (!(React.isValidElement(Stop.Component) || typeof Stop.Component === 'function')) {
      throw new Error(
        `\`createCanal\` could not find a valid \`Component\` key for argument ${index +
          1}. Received: ${JSON.stringify(Stop)}`
      );
    }
  }

  @observer
  class CanalComponent extends ReactComponent<CanalComponentProps<V>> {
    static defaultProps = {
      style: StyleSheet.absoluteFill,
    };

    canal = new Canal(StopsList);

    constructor(props: CanalComponentProps<V>) {
      super(props);
      const navigation = Navigation.getInstance();
      navigation.canalsSubject.next(this.canal);
    }

    render() {
      return (
        <View style={this.props.style}>
          {Navigation.getInstance().state[this.canal.id].map(({ Component, name }) => (
            <View style={StyleSheet.absoluteFill} key={name}>
              <Component />
            </View>
          ))}
        </View>
      );
    }
  }

  // @ts-ignore
  return CanalComponent;
};
