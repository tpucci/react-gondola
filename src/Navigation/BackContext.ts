import { createContext } from 'react';
import { Observable } from 'rxjs';
import { IBackEvent } from './BackHandlerDelegate';
import { Navigation } from './Navigation';

interface IBackContext {
  back$: Observable<IBackEvent>;
}

export const BackContext = createContext<IBackContext>(
  Navigation.instance.backHandlerDelegate.defaultBackContext
);
