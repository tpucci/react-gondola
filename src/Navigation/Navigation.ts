import { FullScreenDelegate } from './FullScreenDelegate';
import { BackHandlerDelegate } from './BackHandlerDelegate';

export class Navigation {
  static get instance() {
    if (Navigation.singleton) {
      return Navigation.singleton;
    }
    Navigation.singleton = new Navigation();
    return Navigation.singleton;
  }
  private static singleton: Navigation;

  fullScreenDelegate = new FullScreenDelegate();

  backHandlerDelegate = new BackHandlerDelegate();

  private constructor() {}
}
