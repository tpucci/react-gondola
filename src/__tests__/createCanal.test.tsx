import React from 'react';
import { View } from 'react-native';
import TestRenderer from 'react-test-renderer';

import { createCanal } from '../createCanal';
import { Navigation } from '../Navigation.store';
import { Canal } from '../Canal';

describe('createCanal', () => {
  it('throws an error if first arg is not a Component', () => {
    try {
      // @ts-ignore
      createCanal({ name: 'a', Component: 'aaa' });
    } catch (error) {
      expect(error.message).toBe(
        '`createCanal` could not find a valid `Component` key for argument 1. Received: {"name":"a","Component":"aaa"}'
      );
    }
    expect.assertions(1);
  });

  it('renders the first page when mounted', () => {
    const Transitioner = createCanal({ name: 'a', Component: View });
    const testRenderer = TestRenderer.create(<Transitioner />);
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  it('throws an error if any arg is not a Component', () => {
    try {
      // @ts-ignore
      createCanal({ name: 'a', Component: View }, { name: 'b', Component: 'aaa' });
    } catch (error) {
      expect(error.message).toBe(
        '`createCanal` could not find a valid `Component` key for argument 2. Received: {"name":"b","Component":"aaa"}'
      );
    }
    expect.assertions(1);
  });

  it('emits the new canal to the navigation store', () => {
    const PageOne = () => <View />;
    const PageTwo = () => <View />;
    const Transitioner = createCanal(
      { name: 'pageOne', Component: PageOne },
      { name: 'pageTwo', Component: PageTwo }
    );
    const expectedCanal = new Canal([
      { name: 'pageOne', Component: PageOne },
      { name: 'pageTwo', Component: PageTwo },
    ]);
    Navigation.getInstance().canalsSubject.subscribe({
      next: canal => {
        expect(canal).toEqual(expectedCanal);
      },
    });
    TestRenderer.create(<Transitioner />);
    expect.assertions(1);
  });

  it('throws an error if name is missing', () => {
    try {
      // @ts-ignore
      createCanal({ Component: View });
    } catch (error) {
      expect(error.message).toBe(
        '`createCanal` could not find a valid `name` key for argument 1. Received: {}'
      );
    }
    expect.assertions(1);
  });

  // @TODO 19-08-01 Pass this test
  xit('throws an error some names are duplicated', () => {
    try {
      // @ts-ignore
      createCanal({ name: 'a', Component: View }, { name: 'a', Component: View });
    } catch (error) {
      expect(error.message).toBe('`createCanal` found duplicated `name: a` key.');
    }
    expect.assertions(1);
  });
});
