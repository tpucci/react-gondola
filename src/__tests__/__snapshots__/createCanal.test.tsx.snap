// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCanal renders the first page when mounted 1`] = `
<View
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View />
  </View>
</View>
`;
